framework/src/meshgenerators/CutMeshByPlaneGenerator.C

```diff
@@ -0,0 +1,458 @@
+  elem_side_normal_list.resize(4);
+  for (unsigned int i = 0; i < 4; i++)
+  {
+    auto elem_side = mesh.elem_ptr(elem_id)->side_ptr(i);
+    elem_side_normal_list[i] = (*elem_side->node_ptr(1) - *elem_side->node_ptr(0))
+                                   .cross(*elem_side->node_ptr(2) - *elem_side->node_ptr(1))
+                                   .unit();
+  }
+
+  std::vector<std::vector<boundary_id_type>> elem_side_list;
+  elem_side_list.resize(4);
+  for (unsigned int i = 0; i < bdry_side_list.size(); i++)
+  {
+    if (std::get<0>(bdry_side_list[i]) == elem_id)
+    {
+      elem_side_list[std::get<1>(bdry_side_list[i])].push_back(std::get<2>(bdry_side_list[i]));
+    }
+  }
+
+  std::vector<PointPlaneRelationIndex> node_plane_relation(4);
+  std::vector<Node *> tet4_nodes(4);
+  std::vector<Node *> tet4_nodes_on_plane;
+  std::vector<Node *> tet4_nodes_outside_plane;
+  std::vector<Node *> tet4_nodes_inside_plane;
+  for (unsigned int i = 0; i < 4; i++)
+  {
+    tet4_nodes[i] = mesh.elem_ptr(elem_id)->node_ptr(i);
+    node_plane_relation[i] = pointPlaneRelation(*tet4_nodes[i], plane_point, plane_normal);
+    if (node_plane_relation[i] == PointPlaneRelationIndex::on_plane)
+      tet4_nodes_on_plane.push_back(tet4_nodes[i]);
+    else if (node_plane_relation[i] == PointPlaneRelationIndex::plane_normal_side)
+      tet4_nodes_outside_plane.push_back(tet4_nodes[i]);
+    else
+      tet4_nodes_inside_plane.push_back(tet4_nodes[i]);
+  }
+  std::vector<Elem *> elems_Tet4;
+  bool new_elements_created(false);
+  // No cutting needed if no nodes are outside the plane
+  if (tet4_nodes_outside_plane.size() == 0)
+  {
+    if (tet4_nodes_on_plane.size() == 3)
+    {
+      // Record the element for future cross section boundary assignment
+      elems_Tet4.push_back(mesh.elem_ptr(elem_id));
+    }
+  }
+  // Remove the element if all the nodes are outside the plane
+  else if (tet4_nodes_inside_plane.size() == 0)
+  {
+    mesh.elem_ptr(elem_id)->subdomain_id() = block_id_to_remove;
+    if (tet4_nodes_on_plane.size() == 3)
+    {
+      // I think the neighboring element will be handled,
+      // So we do not need to assign the cross section boundary here
+      // Need recheck
+    }
+  }
+  // As we have nodes on both sides, five different scenarios are possible
+  else
+  {
+    new_elements_created = true;
+    if (tet4_nodes_inside_plane.size() == 1 && tet4_nodes_outside_plane.size() == 3)
+    {
+      std::vector<Node *> new_plane_nodes;
+      // A smaller TET4 element is created, this solution is unique
+      for (const auto & tet4_node_outside_plane : tet4_nodes_outside_plane)
+      {
+        new_plane_nodes.push_back(nonDuplicateNodeCreator(
+            mesh,
+            new_on_plane_nodes,
+            pointPairPlaneInterception(
+                *tet4_node_outside_plane, *tet4_nodes_inside_plane[0], plane_point, plane_normal)));
+      }
+      elems_Tet4.push_back(mesh.add_elem(new Tet4));
```

Comment:
use this to avoid working with raw pointers
```
            new_elem = std::make_unique<Quad4>();
```
like in advancedExtrudgerGenerator